runtime/capsules/src/mctp/recv.rs

```diff
@@ -0,0 +1,232 @@
+// Licensed under the Apache-2.0 license
+
+use crate::mctp::base_protocol::MCTPHeader;
+
+use core::fmt::Write;
+use romtime::println;
+
+use core::cell::Cell;
+
+use kernel::collections::list::{ListLink, ListNode};
+use kernel::utilities::cells::{MapCell, OptionalCell, TakeCell};
+
+use super::base_protocol::{MessageType, MCTP_HDR_SIZE, MCTP_TAG_MASK, MCTP_TAG_OWNER};
+
+/// This trait is implemented to get notified of the messages received
+/// on corresponding message_type.
+pub trait MCTPRxClient {
+    fn receive(&self, dst_eid: u8, msg_type: u8, msg_tag: u8, msg_payload: &[u8], msg_len: usize);
+}
+
+/// Receive state
+pub struct MCTPRxState<'a> {
+    /// Message assembly context
+    msg_terminus: MapCell<MsgTerminus>,
+    /// Expected message types
+    msg_types: Cell<&'static [MessageType]>,
+    /// Client (implements the MCTPRxClient trait)
+    client: OptionalCell<&'a dyn MCTPRxClient>,
+    /// Message buffer
+    msg_payload: TakeCell<'static, [u8]>,
+    /// next MCTPRxState node
+    next: ListLink<'a, MCTPRxState<'a>>,
+}
+
+impl<'a> ListNode<'a, MCTPRxState<'a>> for MCTPRxState<'a> {
+    fn next(&'a self) -> &'a ListLink<'a, MCTPRxState<'a>> {
+        &self.next
+    }
+}
+
+struct MsgTerminus {
+    msg_type: u8,
+    msg_tag: u8,
+    source_eid: u8,
+    tag_owner: u8,
+    start_payload_len: usize,
+    pkt_seq: u8,
+    msg_size: usize,
+}
+
+impl<'a> MCTPRxState<'a> {
+    pub fn new(
+        rx_msg_buf: &'static mut [u8],
+        message_types: &'static [MessageType],
+    ) -> MCTPRxState<'static> {
+        MCTPRxState {
+            msg_terminus: MapCell::empty(),
+            msg_types: Cell::new(message_types),
+            client: OptionalCell::empty(),
+            msg_payload: TakeCell::new(rx_msg_buf),
+            next: ListLink::empty(),
+        }
+    }
+
+    pub fn set_client(&self, client: &'a dyn MCTPRxClient) {
+        self.client.set(client);
+    }
+
+    /// Checks if a message of the given type is expected to be received.
+    ///
+    /// # Arguments
+    /// 'msg_type' - The message type to check if it is expected.
+    ///
+    /// # Returns
+    /// True if the message type is expected, false otherwise.
+    pub fn is_receive_expected(&self, msg_type: MessageType) -> bool {
+        let msg_types = self.msg_types.get();
```

Comment:
I can't decide if this is easier to read, or if the functional style is:

```rust
self.msg_types.get().iter().find(|t| *t == msg_type).unwrap_or_default()
```

Reply:
Changed it to `self.msg_types.get().iter().any(|&t| t == msg_type as u8)`

Reply:
Changed to   ` self.msg_types.get().iter().any(|&exp_type| exp_type == msg_type)`